src/recovery.rs

```diff
@@ -277,12 +283,30 @@ impl Recovery {
             return Ok(());
         }
 
+        if largest_newly_acked_pkt_num == largest_acked && has_ack_eliciting {
+            let latest_rtt = now - largest_newly_acked_sent_time;
+
+            let ack_delay = if epoch == packet::EPOCH_APPLICATION {
+                Duration::from_micros(ack_delay)
+            } else {
+                Duration::from_micros(0)
+            };
+
+            self.update_rtt(latest_rtt, ack_delay, now);
+        }
+
+        // Detect and mark lost packets without removing them from the sent
+        // packets list.
         self.detect_lost_packets(epoch, now, trace_id);
 
+        self.on_packets_acked(acked_bytes, largest_newly_acked_sent_time, now);
```

Comment:
So for me it doesn't match to current recovery pseudocode

    OnPacketsAcked(newly_acked_packets)

Because here we only pass the information of how many bytes acked and sent time of latest packet.

`newly_acked_packets` is a list of acked packet and later it is used to iterate over as follows:

https://github.com/quicwg/base-drafts/blob/master/draft-ietf-quic-recovery.md#on-packet-acknowledgement

Also note that current quiche CC code assumes `on_packet_acked` hook is called for each acked packet, not only for the last one.

This will make a difference because CC state can be changed per acked packet.

Reply:
For example newly_acked_packet has 5 packets, each ack'ing 1000 bytes, starting with cwnd = 1000 ssthresh = 3000 and assume Reno.
If we follow the current code, on_packets_acked(5000, <time>, now) will result in
cwnd = 6000 ssthresh = 3000
But in the current code, CC will switch to congestion avoidance when cwnd reaches to 3000, so the result will be
cwnd = 3908 ssthresh = 3000

Reply:
Yes, you are right, I missed that part. Should be fixed now, though one of the CUBIC tests now fails, will need to investigate. Though if I understand RFC8312 right, the CUBIC congestion avoidance should only be invoked once per-ACK, rather than for each acked packet?

Reply:
CUBIC (or reno or other cc) cwnd update is per acked packet in QUIC. I don't understand why each CC code need a change. I think simply following the pseudo code of on_packets_acked() (iterating over acked packet.. and we already have on_packet_acked() per packet so simply calling it multiple times) will be good enough for now?

Reply:
Because in that case we need to create temporary vectors with the acked packets which is a potential DoS vector.

In any case I changed the code to do just that now. We can reconsider this later on.